### PR TITLE
Support time based retention

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary

Fixes: #17082

Changes

- Default tiers enabled 3
   - Tier 0 - high resolution (update every)
   - Tier 1 - per minute
   - Tier 2 - per hour
   - Configure disk space using `dbengine tier x disk space MB = nnnn`
     - eg For tier 0,  use `dbengine tier 0 disk space MB = 1024` 
   - Configure time retention using `dbengine tier x retention days = nnnn`
     - eg For tier 0,  use `dbengine tier 0 retention days = 14` 
- Backfill option is now global for all tiers (none, full, new)
   - option `dbengine tier backfill`
- `dbengine tier x disk space MB` options can be 0 to use as much space needed for the desired time based retention
   - The remaining disk space for each tier (minus 5%) will be used for the dbengine retention chart (and alert)

The disk space related options 
- dbengine multihost disk space MB
- dbengine disk space MB

have been removed. The `dbengine multihost disk space MB` is automatically renamed to `dbengine tier 0 disk space MB`

Similarly to configure db space for other tiers (up to 5) eg
- dbengine tier 0 disk space MB
- dbengine tier 1 disk space MB
- dbengine tier 2 disk space MB
- dbengine tier 3 disk space MB
- dbengine tier 4 disk space MB

You can disable disk space checks (use available space) by specifying 0 as disk space eg.
- dbengine tier 0 disk space MB = 0

### Defaults
The default settings on a newly installed agent when no options are set are:
- High resolution tier (tier 0)
  - Default diskspace 1GB or 14 days of data
  - Store values per update every (default = 1)
-  Tier 1
   - Default diskspace 1GB or 90 days of data
   - Store values per minute
-  Tier 2
   - Default diskspace 1GB or 2 years (2 x 365 days) of data
   - Store values per hour
- Retention charts that store the percentage of space and time used (vs the configured values)

### Time retention not automatically activated
If the following options have been specified in the config (`[db]` section) the retention days will not be
automatically activated. In this case the `dbengine tier x retention days` option must be used.

- dbengine multihost disk space MB
- dbengine tier 1 update every iterations
- dbengine tier 2 update every iterations
- dbengine tier 3 update every iterations
- dbengine tier 4 update every iterations
- dbengine tier 1 disk space MB 
  - or dbengine tier 1 multihost disk space MB
- dbengine tier 2 disk space MB 
  - or dbengine tier 2 multihost disk space MB
- dbengine tier 3 disk space MB
  - or dbengine tier 3 multihost disk space MB
- dbengine tier 4 disk space MB 
  - or dbengine tier 4 multihost disk space MB

### CODE REVIEW

File: src/daemon/main.c [hunk 1]
```diff
@@ -1093,7 +1093,23 @@ static int get_hostname(char *buf, size_t buf_size) {
     return gethostname(buf, buf_size);
 }
 
-static void get_netdata_configured_variables() {
+static int get_valid_update_every(int update_every) {
```
Comment (thiagoftsm) [inline src/daemon/main.c:1096]: Considering the name that function has, should we write an `info` when the input is different of expected values/?

Comment (thiagoftsm) [reply src/daemon/main.c:1096]: I saw you are raising an alert outside the function.

**Comment by thiagoftsm** [comment]
@stelfrag I ran different tests with this PR, and I did not observe anything anomalous.

File: src/database/engine/rrdengine.c [hunk 2]
```diff
@@ -1688,6 +1762,73 @@ static inline void worker_dispatch_query_prep(struct rrdeng_cmd cmd, bool from_w
...
+    for (size_t tier = 0; tier < storage_tiers; tier++) {
+        if (init == false) {
+            char id[200];
+            snprintfz(id, sizeof(id) - 1, "dbengine_retention_tier%zu", tier);
+            stats[tier].st = rrdset_create_localhost(
+                "netdata",
+                id,
+                NULL,
+                "dbengine",
+                "dbengine_retention",
```
Comment (ilyam8) [inline src/database/engine/rrdengine.c:1789]:
```suggestion
                "netdata.dbengine_retention",
```

Comment (ilyam8) [reply src/database/engine/rrdengine.c:1789]: And please add "dbengine_tier" label.

File: src/database/engine/rrdengine.c [hunk 3]
```diff
@@ -1688,6 +1762,75 @@ static inline void worker_dispatch_query_prep(struct rrdeng_cmd cmd, bool from_w
...
+                200000,
+                10,
+                RRDSET_TYPE_LINE);
+
+            stats[tier].rd_space = rrddim_add(stats[tier].st, "space", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+            stats[tier].rd_time = rrddim_add(stats[tier].st, "time", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+
+            char tier_str[5];
+            snprintfz(tier_str, 4, "%ld", tier);
+            rrdlabels_add(stats[tier].st->rrdlabels, "dbengine_tier", tier_str, RRDLABEL_SRC_AUTO);
```
Comment (ilyam8) [inline src/database/engine/rrdengine.c:1803]:
```suggestion
            rrdlabels_add(stats[tier].st->rrdlabels, "tier", tier_str, RRDLABEL_SRC_AUTO);
```


I know that I suggested "dbengine_tier"...changing to just "tier" 😄

File: src/database/engine/rrdengine.c [hunk 4]
```diff
@@ -1688,6 +1762,106 @@ static inline void worker_dispatch_query_prep(struct rrdeng_cmd cmd, bool from_w
...
+                "stats",
+                200000,
+                10,
+                RRDSET_TYPE_LINE);
+
+            stats[tier].rd_space = rrddim_add(stats[tier].st, "space", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+            stats[tier].rd_time = rrddim_add(stats[tier].st, "time", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
+
+            char tier_str[5];
+            snprintfz(tier_str, 4, "%ld", tier);
```
Comment (thiagoftsm) [inline src/database/engine/rrdengine.c:1824]: During compilation, I have this warning:

```sh
/home/thiago/Netdata/tests_netdata/src/database/engine/rrdengine.c:1824:39: warning: format '%ld' expects argument of type 'long int', but argument 4 has type 'size_t' {aka 'long unsigned int'} [-Wformat=]
 1824 |             snprintfz(tier_str, 4, "%ld", tier);
      |                                     ~~^   ~~~~
      |                                       |   |
      |                                       |   size_t {aka long unsigned int}
```

File: src/database/engine/journalfile.c [hunk 5]
```diff
@@ -669,6 +669,7 @@ static void journalfile_restore_extent_metadata(struct rrdengine_instance *ctx,
     }
 
     time_t now_s = max_acceptable_collected_time();
```
Comment (thiagoftsm) [inline src/database/engine/journalfile.c:671]: I am going to put some warnings that I am receiving for this file, but they do not have relationship with commits:

```sh
/home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c: In function 'journalfile_create':
/home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c:595:30: warning: 'superblock' may be used uninitialized [-Wmaybe-uninitialized]
  595 |     (void) strncpy(superblock->version, RRDENG_JF_VER, RRDENG_VER_SZ);
      |                    ~~~~~~~~~~^~~~~~~~~
/home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c:576:26: note: 'superblock' was declared here
  576 |     struct rrdeng_jf_sb *superblock;
      |                          ^~~~~~~~~~
In function 'journalfile_check_superblock',
    inlined from 'journalfile_load' at /home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c:1552:11:
/home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c:630:11: warning: 'superblock' may be used uninitialized [-Wmaybe-uninitialized]
  630 |     iov = uv_buf_init((void *)superblock, sizeof(*superblock));
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c: In function 'journalfile_load':
/home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c:622:26: note: 'superblock' was declared here
  622 |     struct rrdeng_jf_sb *superblock;
      |                          ^~~~~~~~~~
In function 'journalfile_iterate_transactions',
    inlined from 'journalfile_load' at /home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c:1562:14:
/home/thiago/Netdata/tests_netdata/src/database/engine/journalfile.c:828:15: warning: 'buf' may be used uninitialized [-Wmaybe-uninitialized]
  828 |         iov = uv_buf_init(buf, size_bytes);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: src/database/engine/datafile.c [hunk 6]
```diff
@@ -543,7 +543,7 @@ int init_data_files(struct rrdengine_instance *ctx)
         if (ctx->loading.create_new_datafile_pair)
             create_new_datafile_pair(ctx, false);
```
Comment (thiagoftsm) [inline src/database/engine/datafile.c:545]: Like comment below:

```sh
/home/thiago/Netdata/tests_netdata/src/database/engine/datafile.c:302:11: warning: 'superblock' may be used uninitialized [-Wmaybe-uninitialized]
  302 |     iov = uv_buf_init((void *)superblock, sizeof(*superblock));
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/thiago/Netdata/tests_netdata/src/database/engine/datafile.c: In function 'scan_data_files':
/home/thiago/Netdata/tests_netdata/src/database/engine/datafile.c:294:26: note: 'superblock' was declared here
  294 |     struct rrdeng_df_sb *superblock;
      |                          ^~~~~~~~~~
/home/thiago/Netdata/tests_netdata/src/database/engine/datafile.c: In function 'create_data_file':
/home/thiago/Netdata/tests_netdata/src/database/engine/datafile.c:267:30: warning: 'superblock' may be used uninitialized [-Wmaybe-uninitialized]
  267 |     (void) strncpy(superblock->version, RRDENG_DF_VER, RRDENG_VER_SZ);
      |                    ~~~~~~~~~~^~~~~~~~~
/home/thiago/Netdata/tests_netdata/src/database/engine/datafile.c:248:26: note: 'superblock' was declared here
  248 |     struct rrdeng_df_sb *superblock;
      |                          ^~~~~~~~~~
```

File: src/database/engine/journalfile.c (hunk 5 above)
Comment (stelfrag) [reply src/database/engine/journalfile.c:671]: I will look at those in a different PR

File: src/database/engine/rrdengine.c (hunk 4 above)
Comment (stelfrag) [reply src/database/engine/rrdengine.c]: resolved

**Review by ilyam8 — APPROVED** [review]

**Pull request merged**